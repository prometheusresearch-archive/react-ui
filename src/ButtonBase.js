/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import * as React from 'react';
import {isString} from 'lodash';

import {create, css, component} from './stylesheet';

let stylesheet = create({
  Root: component.button({
    cursor: css.cursor.pointer,
    textAlign: css.textAlign.left,
    userSelect: css.none,
    WebkitUserSelect: css.none,
    whiteSpace: 'nowrap',
    overflow: 'hidden',
  }),
  Caption: component.div({
    display: css.display.inlineBlock,
  }),
  Icon: 'span',
  IconWrapper: component.div({
    display: css.display.inlineBlock,
  }),
});

/**
 * Button component.
 *
 * Button is clickable element with optional icon and/or caption.
 */
export default class ButtonBase extends React.Component {

  static propTypes = {
    /**
     * If button should be rendered as being pressed.
     */
    active: React.PropTypes.bool,

    /**
     * Button size.
     */
    size: React.PropTypes.oneOf(['small', 'normal', 'large']),

    /**
     * Button's icon.
     */
    icon: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.node
    ]),

    /**
     * Button's caption.
     */
    children: React.PropTypes.node,
  };

  static defaultProps = {
    size: 'normal',
    attach: {},
    stylesheet,
  };

  render() {
    let {
      children, icon,
      disabled, active, size,
      attach,
      variant,
      stylesheet: {Root, Caption, Icon, IconWrapper},
      ...props
    } = this.props;
    variant = {
      active, disabled,
      attachLeft: attach.left,
      attachRight: attach.right,
      attachTop: attach.top,
      attachBottom: attach.bottom,
      small: size === 'small',
      normal: size === 'normal',
      large: size === 'large',
      ...variant,
    };
    if (isString(icon)) {
      icon = <Icon name={icon} />;
    }
    let caption = null;
    if (children) {
      caption = <Caption>{children}</Caption>;
    }
    return (
      <Root
        {...props}
        disabled={disabled}
        variant={variant}
        aria-pressed={active}
        role="button">
        {icon ?
          <IconWrapper variant={{hasCaption: !!children}}>
            {icon}
          </IconWrapper> :
          null}
        {caption}
      </Root>
    );
  }
}

