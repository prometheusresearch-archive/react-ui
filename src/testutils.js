/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {isEqual} from 'lodash/lang';
import indent from 'indent-string';

import {
  findWithType as _findWithType,
  findAllWithType as _findAllWithType,
  getMountedInstance,
  findAll
} from 'react-shallow-testutils';

import {
  createRenderer as _createRenderer
} from 'react-addons-test-utils';

import _inspectElement, {
  inspectReactProp as _inspectReactProp
} from 'inspect-react-element';


import assert from 'power-assert';
import {stub, spy, mock} from 'sinon';

export {stub, spy, mock, assert};

class FormDataMock {

  append = spy();
}

let originalFormData = global.FormData;

export function mockFormData() {
  global.FormData = FormDataMock;
}

export function unmockFormData() {
  global.FormData = originalFormData;
}

function inspectReactProps(node) {
  return '__originalConfig' in node.props ?
    {...node.props.__originalConfig, children: node.props.children} :
    node.props || {};
}

function inspectReactProp(propName, propValue) {
  if (propName === '__originalConfig' || propName === 'key' || propName === 'ref') {
    return null;
  }
  return _inspectReactProp(propName, propValue);
}

export function inspectElement(node) {
  return _inspectElement(node, {inspectReactProps, inspectReactProp});
}

class ReactShallowRenderer {

  constructor() {
    this._renderer = _createRenderer();
  }

  render(element, context) {
    return this._renderer.render(element, context);
  }

  get instance() {
    return getMountedInstance(this._renderer);
  }

  get element() {
    return this._renderer.getRenderOutput();
  }

  findWithType(type) {
    return findWithType(this.element, type);
  }

  findAllWithType(type) {
    return findAllWithType(this.element, type);
  }

  assertElementWithType(type) {
    this.findWithType(type);
  }

  assertNoElementWithType(type) {
    assert(this.findAllWithType(type).length === 0);
  }


  findWithTypeProps(type, props) {
    return findWithTypeProps(this.element, type, props);
  }

  findAllWithTypeProps(type, props) {
    return findAllWithTypeProps(this.element, type, props);
  }

  assertElementWithTypeProps(type, props) {
    this.findWithTypeProps(type, props);
  }

  assertNoElementWithTypeProps(type, props) {
    assert(this.findAllWithTypeProps(type, props).length === 0);
  }


  findWithElement(element) {
    return findWithElement(this.element, element);
  }

  findAllWithElement(element) {
    return findAllWithElement(this.element, element);
  }

  assertElement(element) {
    this.findWithElement(element);
  }

  assertNoElement(element) {
    assert(this.findAllWithElement(element).length === 0);
  }

}

export function createRenderer() {
  return new ReactShallowRenderer();
}

export function findWithType(root, type) {
  let elements = findAll(root, _makeMatchWithType(type));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one partial match for element:
${indent(inspectElement(React.createElement(type)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithType(root, type) {
  return findAll(root, _makeMatchWithType(type));
}

function _makeMatchWithType(type) {
  return function _matchWithType(element) {
    return element && element.type && (
      element.type === type ||
      element.type.Component === type
    );
  };
}

export function findWithTypeProps(root, type, props) {
  let elements = findAll(root, _makeMatchWithElement(type, props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(React.createElement(type, props)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithTypeProps(root, type, props) {
  return findAll(root, _makeMatchWithElement(type, props));
}

export function findWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  let elements = findAll(root, _makeMatchWithElement(spec.type, spec.props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(spec), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  return findAll(root, _makeMatchWithElement(spec.type, spec.props));
}

function _makeMatchWithElement(type, props) {
  let matchWithType = _makeMatchWithType(type);

  let matchWithProps = function(element) {
    for (let key in props) {
      if (!props.hasOwnProperty(key)) {
        continue;
      }
      if (!element.props.hasOwnProperty(key)) {
        continue;
      }
      if (key === '__originalConfig') {
        continue;
      }
      if (!isEqual(element.props[key], props[key])) {
        return false;
      }
    }
    return true;
  };
  return function _matchWithElement(element) {
    return element && matchWithType(element) && matchWithProps(element);
  };
}

function _normalizeReactElementSpec(spec) {
  let {props: {__originalConfig = null}} = spec;
  if (__originalConfig) {
    return {...spec, props: __originalConfig};
  } else {
    return spec;
  }
}

let _createElement = React.createElement;

function createElement(type, config, ...children) {
  config = {...config, __originalConfig: config || {}};
  let element = _createElement(type, config, ...children);
  return element;
}

React.createElement = createElement;

