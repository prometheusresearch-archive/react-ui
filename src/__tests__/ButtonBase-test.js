/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import * as React from 'react';
import TestUtils from 'react-addons-test-utils';
import {assert} from '../testutils';
import ButtonBase from '../ButtonBase';

describe('<ButtonBase />', function() {

  it('renders with just caption', function() {
    let renderer = TestUtils.createRenderer();
    renderer.render(
      <ButtonBase>OK</ButtonBase>
    );
    let root = renderer.getRenderOutput();
    let [iconWrapper, caption] = root.props.children;
    assert(iconWrapper === null);
    assert(caption.props.children === 'OK');
  });

  it('renders with just icon (text)', function() {
    let renderer = TestUtils.createRenderer();
    renderer.render(
      <ButtonBase icon="plus" />
    );
    let root = renderer.getRenderOutput();
    let [{props: {variant: iconVariant, children: icon}}, caption] = root.props.children;
    assert(caption === null);
    assert(icon);
    assert(icon.props.name === 'plus');
    assert(!iconVariant.hasCaption);
  });

  it('renders with just icon (element)', function() {
    let renderer = TestUtils.createRenderer();
    renderer.render(
      <ButtonBase icon={<icon />} />
    );
    let root = renderer.getRenderOutput();
    let [{props: {variant: iconVariant, children: icon}}, caption] = root.props.children;
    assert(caption === null);
    assert(icon);
    assert(icon.type === 'icon');
    assert(!iconVariant.hasCaption);
  });

  it('renders with icon and caption', function() {
    let renderer = TestUtils.createRenderer();
    renderer.render(
      <ButtonBase icon="plus">Caption</ButtonBase>
    );
    let root = renderer.getRenderOutput();
    let [{props: {variant: iconVariant, children: icon}}, caption] = root.props.children;
    assert(icon);
    assert(icon.props.name === 'plus');
    assert(iconVariant.hasCaption);
    assert(caption.props.children === 'Caption');
  });
});

