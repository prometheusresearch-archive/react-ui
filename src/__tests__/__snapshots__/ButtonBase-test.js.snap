// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with icon and caption 1`] = `
<button
  aria-pressed={undefined}
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "display": "inline-block",
        "overflow": "hidden",
        "textAlign": "left",
        "userSelect": "none",
        "verticalAlign": "bottom",
        "whiteSpace": "nowrap",
      },
    }
  }
  disabled={undefined}
  href={undefined}
  role="button"
  style={
    Object {
      "height": undefined,
      "textAlign": undefined,
      "width": undefined,
    }
  }
>
  <div
    className={
      CompiledCSSWithVariants {
        "base": Object {
          "display": "inline-block",
        },
      }
    }
    style={
      Object {
        "marginLeft": 0,
        "marginRight": 4,
      }
    }
  >
    plus
  </div>
  <div
    className={
      CompiledCSSWithVariants {
        "base": Object {
          "display": "inline-block",
        },
      }
    }
  >
    Caption
  </div>
</button>
`;

exports[`renders with just caption 1`] = `
<button
  aria-pressed={undefined}
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "display": "inline-block",
        "overflow": "hidden",
        "textAlign": "left",
        "userSelect": "none",
        "verticalAlign": "bottom",
        "whiteSpace": "nowrap",
      },
    }
  }
  disabled={undefined}
  href={undefined}
  role="button"
  style={
    Object {
      "height": undefined,
      "textAlign": undefined,
      "width": undefined,
    }
  }
>
  <div
    className={
      CompiledCSSWithVariants {
        "base": Object {
          "display": "inline-block",
        },
      }
    }
  >
    OK
  </div>
</button>
`;

exports[`renders with just icon (element) 1`] = `
<button
  aria-pressed={undefined}
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "display": "inline-block",
        "overflow": "hidden",
        "textAlign": "left",
        "userSelect": "none",
        "verticalAlign": "bottom",
        "whiteSpace": "nowrap",
      },
    }
  }
  disabled={undefined}
  href={undefined}
  role="button"
  style={
    Object {
      "height": undefined,
      "textAlign": undefined,
      "width": undefined,
    }
  }
>
  <div
    className={
      CompiledCSSWithVariants {
        "base": Object {
          "display": "inline-block",
        },
      }
    }
    style={
      Object {
        "marginLeft": 0,
        "marginRight": 0,
      }
    }
  >
    <icon />
  </div>
</button>
`;

exports[`renders with just icon (text) 1`] = `
<button
  aria-pressed={undefined}
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "display": "inline-block",
        "overflow": "hidden",
        "textAlign": "left",
        "userSelect": "none",
        "verticalAlign": "bottom",
        "whiteSpace": "nowrap",
      },
    }
  }
  disabled={undefined}
  href={undefined}
  role="button"
  style={
    Object {
      "height": undefined,
      "textAlign": undefined,
      "width": undefined,
    }
  }
>
  <div
    className={
      CompiledCSSWithVariants {
        "base": Object {
          "display": "inline-block",
        },
      }
    }
    style={
      Object {
        "marginLeft": 0,
        "marginRight": 0,
      }
    }
  >
    plus
  </div>
</button>
`;
